### ISSUE TITLE
How to processing text to training canary 180m model?

### ISSUE THREAD

**Issue by PhamDangNguyen** [NVIDIA/NeMo#12957]
Hi
Congrats on the work, I just want to try training model and I have a question.

This code is a tutorial for processing a dataset to train a model with fine-tuning canary-180m-flash, refer [script](https://github.com/NVIDIA/NeMo/blob/main/tutorials/asr/Canary_Multitask_Speech_Model.ipynb): 

``
def build_manifest(data_root, manifest_path):

                                       .......
                    transcript = ' '.join(line.strip().split(' ')[1:]).lower()

                    transcript = transcript.strip()

                                       ........

``
Can I get both punctuation and timestamps in results when inferring with pnc=True and timestamp=yes  when training model with this text dataset?

Following the tutorials, I got a manifest where each example is on one line:

{"audio_filepath": "datasets/LibriLight/librispeech_finetuning/1h/2/clean/5778/12761/5778-12761-0005.flac", "duration": 10.945, "text": "but after our long wandering in rugged mountains where so frequently we had met with disappointments and where the crossing of every ridge displayed some unknown lake or river", "target_lang": "en", "source_lang": "en", "pnc": "False"}

Do I need to set "pnc": "true" or add timestamp to each data point?

**Comment by ankitapasad** [comment]
Hi @PhamDangNguyen 

The sample dataset used in the tutorial does not have punctuation or timestamp labels. To train the model for these capabilities, you'll need to use training data with punctuations (and set `"pnc": True`) and word-level timestamps.

An example manifest entry with timestamp labels: `"text": "<|3|> it's <|7|> <|8|> almost <|9|> <|14|> beyond <|20|> <|20|> conjecture <|28|>"`, as described in the tutorial: "3. Training on a new task: A case of timestamp prediction".

Note that in the tutorial we use a small toy dataset, which won't be enough to train the model to learn anything meaningful. Once you have your dataset prepared, there's also an option to start from a tuned Canary-flash checkpoint; see under "Train Canary model from a Canary flash checkpoint (aka fine-tuning)" in the tutorial. 

Hope that helps!

**Comment by PhamDangNguyen** [comment]
Thanks for your help, @ankitapasad !

**Comment by jeremy110** [comment]
@ankitapasad 
Hello, first of all, thank Nemo team for  open-source contributions. I would like to ask a few questions regarding the paper and implementation:

1. In the paper, OWSM is mentioned. A new method, OWLS, has been introduced in this [paper](https://arxiv.org/pdf/2502.10373). Has NeMo conducted any experiments similar to OWLS to evaluate the relationship between the number of training hours and the WER?

I’ve tried training an ASR model with around 1,000 hours of Japanese and Korean data(https://github.com/NVIDIA/NeMo/issues/12845#issuecomment-2774876408), as well as AST models that translate Japanese and Korean into Chinese and English. The results show some effectiveness. I'm now aiming to further train a multilingual ASR and AST system for Chinese, English, Japanese, and Korean. I’d like to know roughly how many hours of training data are needed to achieve WER < 15%.

2. In the new Canary paper, a 2D bucketing strategy combined with OOMptimizer is used. In Figure 4, GPU compute utilization and batch size are compared. In my own tests, I observed a clear improvement in the input-to-padding ratio, but the batch size doesn’t seem to increase. I’d like to ask if this is related to GPU memory limitations (I’m using a 24GB RTX 4090)?

Red line is 2D bucket + OOMptimizer
![Image](https://github.com/user-attachments/assets/ff03d568-cd27-4213-b823-4f2c8822c32c)
![Image](https://github.com/user-attachments/assets/64ae23d4-f2c1-4c4f-9621-8ad801d8e54c)

**Comment by jeremy110** [comment]
@pzelasko Thank you very much for your reply.

1. I checked afterward and realized that I initially set num_buckets to 20, but later changed it to 30 — that was my mistake.

2. With that in mind, I’d like to ask if my output_to_padding_ratio (as shown in the image) looks normal?

![Image](https://github.com/user-attachments/assets/31c465a8-10e2-434f-919a-35746adfb447)

3. Below is how I set up my model.train_ds. I’d like to ask how I should configure the input for estimate_duration_bins_2d.py to match this setup? I’m using Method 2 — '2) "[[path1.json],[path2.json],...]" - any collection of NeMo manifests' — because when I tried Method 3 and set weights, I encountered an error.

```
train_ds:
    use_lhotse: true
    tarred_audio_filepaths: null
    input_cfg:
      - type: group
        weight: 0.5
        tags:
          task: asr
        input_cfg:
          - type: nemo
            manifest_filepath: /ws_sto/manifest/train/asr/ko_ko_asr_train_filter.json
            tags:
              source_lang: ko
              target_lang: ko
          - type: nemo
            manifest_filepath: /ws_sto/manifest/train/asr/ja_ja_asr_train_filter_n500.json
            tags:
              source_lang: ja
              target_lang: ja
      - type: group
        weight: 0.5
        tags:
          task: ast
        input_cfg:
          - type: nemo
            manifest_filepath: /ws_sto/manifest/train/ast/ja_en_ast_train_filter.json
            tags:
              source_lang: ja
              target_lang: en
          - type: nemo
            manifest_filepath: /ws_sto/manifest/train/ast/ja_zh_ast_train_filter.json
            tags:
              source_lang: ja
              target_lang: zh
          - type: nemo
            manifest_filepath: /ws_sto/manifest/train/ast/ko_en_ast_train_filter.json
            tags:
              source_lang: ko
              target_lang: en
          - type: nemo
            manifest_filepath: /ws_sto/manifest/train/ast/ko_zh_ast_train_filter.json
            tags:
              source_lang: ko
              target_lang: zh
    sample_rate: ${model.sample_rate}
    shuffle: true
    num_workers: 4 #8
    batch_size: null
    batch_duration: 360
    quadratic_duration: 15
    use_bucketing: True
    num_buckets: 30
    bucket_buffer_size: 20000
    shuffle_buffer_size: 10000
    min_duration: 1
    max_duration: 20 #
    bucket_duration_bins: [[2.002,15],[2.002,39],[2.49,19],[2.49,40],[2.87,21],[2.87,44],[3.194,23],[3.194,47],[3.486,24],[3.486,50],[3.75,26],[3.75,53],[3.99,27],[3.99,56],[4.219,28],[4.219,58],[4.44,29],[4.44,60],[4.651,30],[4.651,62],[4.87,31],[4.87,64],[5.08,32],[5.08,67],[5.296,34],[5.296,69],[5.519,35],[5.519,70],[5.74,36],[5.74,73],[5.977,37],[5.977,75],[6.22,38],[6.22,76],[6.487,39],[6.487,80],[6.77,40],[6.77,80],[7.074,41],[7.074,84],[7.439,42],[7.439,85],[7.838,44],[7.838,90],[8.29,46],[8.29,92],[8.81,47],[8.81,94],[9.408,49],[9.408,96],[10.109,52],[10.109,96],[10.998,54],[10.998,101],[12.204,57],[12.204,104],[14.201,61],[14.201,110],[20.0,66],[20.0,116]]
    bucket_batch_size: [126,88,101,77,88,70,82,62,74,58,68,54,65,50,60,48,58,46,56,44,54,42,50,40,48,38,46,38,45,36,42,35,42,34,40,32,38,31,37,30,35,28,33,27,31,25,29,24,28,23,25,21,23,20,21,18,18,15,13,11]
```
